packages/nouns-contracts/contracts/Rewards.sol

```diff
@@ -0,0 +1,591 @@
+// SPDX-License-Identifier: GPL-3.0
+
+/*********************************
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░█████████░░█████████░░░ *
+ * ░░░░░░██░░░████░░██░░░████░░░ *
+ * ░░██████░░░████████░░░████░░░ *
+ * ░░██░░██░░░████░░██░░░████░░░ *
+ * ░░██░░██░░░████░░██░░░████░░░ *
+ * ░░░░░░█████████░░█████████░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ *
+ *********************************/
+
+pragma solidity ^0.8.19;
+
+import { INounsDAOLogicV3 } from './interfaces/INounsDAOLogicV3.sol';
+import { INounsAuctionHouseV2 } from './interfaces/INounsAuctionHouseV2.sol';
+import { NounsDAOTypes } from './governance/NounsDAOInterfaces.sol';
+import { NounsClientToken } from './client-incentives/NounsClientToken.sol';
+import { IERC20 } from '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import { UUPSUpgradeable } from '@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol';
+import { SafeERC20 } from '@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol';
+import { PausableUpgradeable } from '@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol';
+import { InMemoryMapping } from './libs/InMemoryMapping.sol';
+
+contract Rewards is NounsClientToken, UUPSUpgradeable, PausableUpgradeable {
+    using SafeERC20 for IERC20;
+    using InMemoryMapping for InMemoryMapping.Mapping;
+
+    /**
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     *   EVENTS
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    event ClientRewarded(uint32 indexed clientId, uint256 amount);
+    event ClientBalanceWithdrawal(uint32 indexed clientId, uint256 amount, address to);
+    event AuctionRewardsUpdated(uint256 firstAuctionId, uint256 lastAuctionId);
+    event ProposalRewardsUpdated(uint32 firstProposalId, uint32 lastProposalId);
+
+    /**
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     *   CONSTANTS
+     * ░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
+     */
+
+    /// @notice The maximum priority fee used to cap gas refunds
+    uint256 public constant MAX_REFUND_PRIORITY_FEE = 2 gwei;
```

Comment:
could be nice to extract gas refund logic to a library with internal functions, so these constants go in there and this contract becomes a bit easier to read. wdyt?

Reply:
@eladmallel I actually started with doing a library because I wanted to reuse logic, but then didn't because this depended on the ERC20 token.
I can and pass the ERC20 token as a parameter.
won't be so reusable, but maybe easier to read

Reply:
done here: bbfe1ecee82bb3291a6c0102df92a86ab79330b6
wdyt?